src/compute/src/server.rs

```diff
@@ -172,14 +219,154 @@ impl<'w, A: Allocate> Worker<'w, A> {
+                    for (dataflows_part, objects_to_build) in
+                        dataflows_parts.iter_mut().zip(builds_parts)
+                    {
+                        dataflows_part.push(DataflowDescription {
+                            source_imports: dataflow.source_imports.clone(),
+                            index_imports: dataflow.index_imports.clone(),
+                            objects_to_build,
+                            index_exports: dataflow.index_exports.clone(),
+                            sink_exports: dataflow.sink_exports.clone(),
+                            as_of: dataflow.as_of.clone(),
+                            until: dataflow.until.clone(),
+                            debug_name: dataflow.debug_name.clone(),
+                        });
+                    }
+                }
+                dataflows_parts
+                    .into_iter()
+                    .map(ComputeCommand::CreateDataflows)
+                    .collect()
+            }
+            command => vec![command; parts],
+        }
+    }
+
+    fn setup_channel_and_run_client(
+        &mut self,
+        command_rx: CommandReceiver,
+        response_tx: ResponseSender,
+        activator_tx: ActivatorSender,
+    ) {
+        let cmd_queue = Rc::new(RefCell::new(
+            VecDeque::<Result<ComputeCommand, TryRecvError>>::new(),
+        ));
+        let peers = self.timely_worker.peers();
+        let idx = self.timely_worker.index();
+
+        {
+            let cmd_queue = Rc::clone(&cmd_queue);
+
+            self.timely_worker.dataflow::<u64, _, _>(move |scope| {
+                source(scope, "CmdSource", |capability, info| {
+                    // Send activator for this operator back
+                    let activator = scope.sync_activator_for(&info.address[..]);
+                    activator_tx.send(activator).expect("activator_tx working");
+
+                    //Hold onto capbility until we receive a disconnected error
+                    let mut cap_opt = Some(capability);
+
+                    move |output| {
+                        let mut disconnected = false;
+                        if let Some(cap) = cap_opt.as_mut() {
+                            let time = cap.time().clone();
+                            let mut session = output.session(&cap);
+
+                            loop {
+                                match command_rx.try_recv() {
+                                    Ok(cmd) => {
+                                        // Commands must never be sent to another worker. This
+                                        // implementation does not guarantee an ordering of events
+                                        // sent to different workers.
+                                        assert!(idx == 0);
+                                        session.give_iterator(
+                                            Self::split_command(cmd, peers).into_iter().enumerate(),
```

Comment:
Is the `enumerate()` here necessary? Looks like you ignore the index on line 346 anyway.

Reply:
It is yes, the `Exchange::new(...)` of the `unary_frontier` is looking at the index

Reply:
Ah, missed that!